compiler/resolution/functionResolution.cpp

```diff
@@ -7392,46 +7412,93 @@ static void resolveAutoCopies() {
-   flag to the parent types as well.
+   flag to the parent types as well. While doing so, we also
+   propagate whether or not that type is not Plain-old-Data
+   (which would mean it does not need auto copies or auto
+    destroys - bit copies will do).
 
-   After this function is called on a type, that type and any field
-   types must have either FLAG_IGNORE_NOINIT or FLAG_ALLOW_NOINIT set.
+   After this function is called on a record type, that type and any field
+   record types must have either FLAG_POD or FLAG_NOT_POD set.
 
-   Returns true if FLAG_IGNORE_NOINIT is set, false otherwise.
+   After setting either FLAG_POD or FLAG_NOT_POD if necessary,
+   returns true if FLAG_NOT_POD is set, false otherwise.
  */
-static bool propagateIgnoreNoinit(Type* t) {
+static bool propagateNotPOD(Type* t) {
   // Move past those records we've already handled
-//  if( t->symbol->hasFlag(FLAG_ALLOW_NOINIT) )
-//    return false;
-  if( t->symbol->hasFlag(FLAG_IGNORE_NOINIT) )
+  if (t->symbol->hasFlag(FLAG_POD))
+    return false;
+  if (t->symbol->hasFlag(FLAG_NOT_POD))
     return true;
+
   // Move past classes and non-aggregate types.
   if( (!isAggregateType(t)) || isClass(t) )
-    return t->symbol->hasFlag(FLAG_IGNORE_NOINIT);
-
+    return false;
 
   AggregateType *at = (AggregateType*) t;
-  bool ignoreNoinit = false;
-
-  // setting the flag here avoids recursion problems.
-//  at->symbol->addFlag(FLAG_ALLOW_NOINIT);
+  bool notPOD = false;
 
   for_fields(field, at) {
     Type* ft = field->typeInfo();
-    ignoreNoinit |= propagateIgnoreNoinit(ft);
-  }
- 
-  if (ignoreNoinit) {
-//    at->symbol->removeFlag(FLAG_ALLOW_NOINIT);
-    at->symbol->addFlag(FLAG_IGNORE_NOINIT);
+    notPOD |= propagateNotPOD(ft);
+  }
+
+  // make sure we have resolve auto copy/auto destroy.
+  resolveAutoCopy(t);
+  resolveAutoDestroy(t);
+
+  // also check for a non-compiler generated
+  // autocopy/autodestroy.
+  FnSymbol* autoCopyFn = autoCopyMap.get(t);
+  FnSymbol* autoDestroyFn = autoDestroyMap.get(t);
+  FnSymbol* destructor = t->destructor;
+
+  // Ignore invisible (compiler-defined) functions.
+  if (autoCopyFn && autoCopyFn->hasFlag(FLAG_COMPILER_GENERATED))
+    autoCopyFn = NULL;
+  if (autoDestroyFn && autoDestroyFn->hasFlag(FLAG_COMPILER_GENERATED))
+    autoDestroyFn = NULL;
+  if (destructor && destructor->hasFlag(FLAG_COMPILER_GENERATED))
+    destructor = NULL;
+
+  // if it has flag ignore no-init, it's not pod
+  // if it has a user-specified auto copy / auto destroy, it's not pod 
+  // if it has a user-specified destructor, it's not pod
+  if( t->symbol->hasFlag(FLAG_IGNORE_NOINIT) ||
+      autoCopyFn || autoDestroyFn ||
+      destructor )
+    notPOD = true;
+
+  if( !notPOD && hasUserAssign(t) )
+    notPOD = true;
+
+  if (notPOD) {
+    at->symbol->addFlag(FLAG_NOT_POD);
+  } else {
+    at->symbol->addFlag(FLAG_POD);
   }
 
-  return ignoreNoinit;
+  return notPOD;
+}
+
```

Comment:
There are subtle differences between isPOD() and propagateNotPOD(). The code should be clear about when it is legal to call isPOD().